Fix our tests on windows

iarna opened the pull request:
Making a PR for this to track our work. 

After cloning the repo make sure you do an `npm rebuild` in order to get cmdshims for your binaries.

Also be aware that `tap`'s glob support doesn't seem to support windows-type-slashes, so `npm run tap test/tap/l*.js` works but `npm run tap test\tap\l*.js` does not.

I've not tested test/tap/registry.js as I don't have couchdb in windows as yet.

Currently failing tests are: (as run from the "Github Shell", which is a powershell with git stuff in the path)
## @iarna
#### subshell issues
- [x] test/tap/builtin-config.js _fails when run from a path with a space in its name_
- [x] test/tap/verify-no-lifecycle-on-repo.js _(opened up a new cmd window‼)_
#### symlinks
- [x] test/tap/do-not-remove-other-bins.js
- [x] test/tap/gently-rm-linked-module.js _fails when run with `npm run tap`
- [x] test/tap/gently-rm-symlinked-global-dir.js
- [x] test/tap/lifecycle-path.js (yes, this one is in $PATH too)
- [x] test/tap/publish-access-unscoped-restricted-fails.js
- [x] test/tap/symlink-cycle.js
- [x] test/tap/unit-gentlyrm.js
#### unicode not defaulting to true
- [x] test/tap/extraneous-dep-cycle-ls-ok.js _fails only because we're not defaulting unicode to true_
- [x] test/tap/ls-l-depth-0.js _fails only because we're not defaulting unicode to true_
#### $PATH
- [x] test/tap/lifecycle-path.js _pmfi?_
#### $ var substitutions
- [x] test/tap/umask-lifecycle.js
#### config/env parsing issues

Usually just stuff with strings ending up as numbers, I think
- [x] test/tap/config-basic.js
- [x] test/tap/config-builtin.js
## @zkat
#### git-related
- [x] test/tap/add-remote-git-file.js
- [x] test/tap/github-shortcut.js
- [x] test/tap/version-update-shrinkwrap.js
- [x] test/tap/version-lifecycle.js
#### simple pathname matches
- [x] test/tap/dedupe-scoped.js
- [x] test/tap/outdated-depth-integer.js
- [x] test/tap/unit-child-path.js
- [x] test/tap/unit-module-name.js
- [x] test/tap/unit-package-id.js
#### maximum callstack exceeded :scream:
- [x] test/tap/legacy-\* _(fail for many reasons, not the least that mkdirp.sync is blowing the stack recursing: there may also be a Tacks bug)_
#### `\r\n` issues
- [x] test/tap/config-certfile.js (there looks like some `\r` chars are missing? in the cert as read?)
- [x] test/tap/graceful-restart.js (this one is a simple issue with `\r\n` in the test checks)
- [x] test/tap/outdated.js
#### Misc
- [x] test/tap/check-permissions.js
- [x] test/tap/config-edit.js _fails if there's a space in your path_
- [x] test/tap/cruft-test.js
- [x] test/tap/install-scoped-link.js _fails if there's a space in your path_
- [x] test/tap/git-cache-no-hooks.js _only fails with `npm t`_
- [x] test/tap/logout.js
- [x] test/tap/outdated-depth-deep.js _fails when run from a path with a space in its name_
## Unassigned
#### ENOSYS kill
- [x] test/tap/publish-config.js
- [x] test/tap/unpublish-config.js
#### misc

These are either unclear problems, or ungrouped ones that don't seem to have other related ones.
- [x] test/tap/config-project.js
- [x] test/tap/git-cache-locking.js
- [x] test/tap/git-npmignore.js
- [x] test/tap/git-races.js _fails if there's a space in your path_
- [x] test/tap/install-bad-man.js
- [x] test/tap/install-link-scripts.js
- [x] test/tap/install-man.js
- [x] test/tap/install-save-local.js
- [x] test/tap/install-scoped-already-installed.js
- [x] test/tap/logout-scoped.js
- [x] test/tap/shrinkwrap-save-with-existing-dev-deps.js
- [x] test/tap/sorted-package-json.js
- [x] test/tap/spawn-enoent-help.js
- [x] test/tap/cache-shasum.js _fails if there's a space in your path_

iarna committed:
test: fix Windows-specific test issues

This commit is a squash of many other commits. They all act on
individual tests. Some of them have notable messages, which are
preserved below:

test: rewrite verify-no-lifecycle-on-repo to work on windows

On windows, the previous version of this would open a new cmd window, run
echo in it, and then leave it there, interactive, and wait for the window to
be closed.

test: Fix gently-rm-linked-module on windows

We just need to make our link explicitly be a junction.  This will work as
long as you run tests on NTFS.  It does NOT work on, for instance, vbox
shared folders.

test: Fix gently-rm-symlinked-global-dir to work on windows

To do this we needed to do three things:
1. Be aware that npm doesn't use the same global path structure on
windows. (No lib, bin, etc, all flat.)
2. Not tie the tests to platform-specific unicode support
3. Create our links as junctions, so ordinary users can make them

test: Rewrite lifecycle-path to use `npm run env`

This test predates the existence of `npm run env`-- by changing to use
it we both get to get rid of an unneeded fixture and fix things to
work on windows.

test: fix config-edit on windows

`child_process.spawn`, which is what the `editor` package uses
does not support running straight-up scripts on Windows. Only `.exe`s

For this reason, it's better to just explicitly use `node` here.

test: Fix publish-access-unscoped-restricted-fails on windows

This involved switching from a direct integration test to using common.npm,
as the former method wasn't resulting in all files being closed which meant
that `rimraf` of the test dir couldn't finish on Windows.

test: Don't Object.keys on non typeof == object

`Object.keys` is valid on any non-null, not undefined value in current
versions of Node.js, but older versions would throw an error for any
non typeof == object.

test: fix gently-rm-linked-module when run from `npm run tap`

`npm run tap` sets the npm_config_prefix env var, which was
overriding the test's own attempt at setting the prefix (via the
PREFIX env var).

test: fix race in cache-shasum by using async rimraf

iarna committed:
src: Refactor windows detection out into lib functions

iarna committed:
explore: Fix running sub-commands on windows

zkat committed:
util: fix module-name on windows

iarna committed:
link: fix absolute link resolution

This previously had win32 specific bits that are unnecessary.  Previously
there are some bugs in how relative links were computed if `cwd` !=
`target`.

jdalton commented:
The WebApps and Frameworks team (my team) on Edge is looking into the symlinks issue this month.
Related to [win-symlink-test](https://github.com/iarna/win-symlink-test).

jdalton commented:
We're investigating if [bash on Windows](https://blogs.windows.com/buildingapps/2016/03/30/run-bash-on-ubuntu-on-windows/) will solve this issue.

jdalton commented:
If not we'll contact the approp folks to get this on the right feature team's radar.

jdalton commented:
\cc @AmazingJaze

othiym23 committed:
ci: add AppVeyor to CI matrix

zkat committed:
normalize-git-url@3.0.2

jdalton commented:
@iarna To help get the right folks at MS looking into this can you explain the impact that non-admin symlink support would have on npm. Is it perf, is it edge case support, etc?

iarna commented:
@jdalton I'll write up something and CC you on Twitter with it.

jdalton commented:
:sunglasses:

othiym23 committed:
test: enable coverage, handle coverage

nyc has come a long way over the last few months, and now only one test
is wily enough to detect its presence. Handle that, and we get coverage
information about the CLI!

othiym23 committed:
git: remove a function call from Git helper

iarna committed:
deps: Eliminate use of npa for dependency resolution

Previously we would match against realize-package-specifier sometimes, npa
other times and in the case of file deps this would (of course) result in
inconsistent results.

This patch does two things, first it replaces all of the npa calls with
realize-package-specifier and makes those functions callback based. Second,
it caches the results from realize-package-specifier so that we don't have
to keep hitting the disk for the same resolution.

zkat committed:
npm-package-arg@4.1.1

Fixes some Windows-related stuff

zkat committed:
realize-package-specifier@3.0.2

iarna committed:
unbuild: Fix unbuilding bins for scoped modules

Previously it was selecting the containing `node_modules` by doing
`path.dirname(modulepath)` which obviously doesn't work for scoped
modules.

iarna committed:
common-tap: Factor out exec path escaping

iarna commented:
```
> tap test/tap/git-dependency-install-link.js
test/tap/git-dependency-install-link.js ............. 11/13 8s
  install from git repo [with --link]
  not ok error reading child package.json
    found:
      name: Error
      stack: "Error: ENOENT: no such file or directory, open 'C:\\Users\\Rebecca\\Documents\\GitHub\\npm\\test\\tap\\git-dependency-install-link\\node_modules\\child\\package.json'\n    at Error (native)"
      message: "ENOENT: no such file or directory, open 'C:\\Users\\Rebecca\\Documents\\GitHub\\npm\\test\\tap\\git-dependency-install-link\\node_modules\\child\\package.json'"
      errno: -4058
      code: ENOENT
      syscall: open
      path: "C:\\Users\\Rebecca\\Documents\\GitHub\\npm\\test\\tap\\git-dependency-install-link\\node_modules\\child\\package.json"
    at:
      line: 92
      column: 9
      file: test/tap/git-dependency-install-link.js
    stack: >
      test/tap/git-dependency-install-link.js:92:9

      ReadFileContext.<anonymous>
      (node_modules/read-package-json/read-json.js:90:21)

      ReadFileContext.callback (node_modules/graceful-fs/graceful-fs.js:78:16)

      FSReqWrap.readFileAfterOpen [as oncomplete] (fs.js:324:13)
```

iarna commented:
```
> tap test/tap/cache-shasum.js
test/tap/cache-shasum.js .............................. 4/5 2s
  cleanup
  not ok Error: ENOTEMPTY: directory not empty, rmdir 'C:\Users\Rebecca\Documents\GitHub\npm\test\tap\cache-shasum\request\2.27.0'
    at:
      native: true
      function: Error
    stack: |
      Object.fs.rmdirSync (fs.js:806:18)
      rmkidsSync (node_modules/rimraf/rimraf.js:334:11)
      rmdirSync (node_modules/rimraf/rimraf.js:324:7)
      rimrafSync (node_modules/rimraf/rimraf.js:295:9)
      node_modules/rimraf/rimraf.js:332:5
      rmkidsSync (node_modules/rimraf/rimraf.js:331:26)
      rmdirSync (node_modules/rimraf/rimraf.js:324:7)
      rimrafSync (node_modules/rimraf/rimraf.js:295:9)
      node_modules/rimraf/rimraf.js:332:5
      rmkidsSync (node_modules/rimraf/rimraf.js:331:26)
      rmdirSync (node_modules/rimraf/rimraf.js:324:7)
      Function.rimrafSync [as sync] (node_modules/rimraf/rimraf.js:295:9)
      test/tap/cache-shasum.js:51:10
    errno: -4051
    code: ENOTEMPTY
    syscall: rmdir
    path: "C:\\Users\\Rebecca\\Documents\\GitHub\\npm\\test\\tap\\cache-shasum\\request\\2.27.0"
    test: cleanup
    message: "Error: ENOTEMPTY: directory not empty, rmdir 'C:\\Users\\Rebecca\\Documents\\GitHub\\npm\\test\\tap\\cache-shasum\\request\\2.27.0'"
```

xialeizhou commented on test/tap/git-npmignore.js:
```diff
@@ -1,5 +1,5 @@
-var cat = require('graceful-fs').writeFileSync
-var exec = require('child_process').exec
+var writeFileSync = require('graceful-fs').writeFileSync
```
cool~

zkat commented:
@othiym23 can I get a due diligence sheep?

othiym23 commented:
To record our in-air resolution: AppVeyor commits squished down to one, bug fixes left independent, try to make the test changes not too chunky but squash them down as well. :sheep::rocket::fire::tada:

othiym23 committed:
src: bring code up to standard@6 compatibility

zkat committed:
standard@6.0.8

iarna commented:
Landed in 3.9.0!!!
